src/core/Internal.ml

```diff
@@ -1661,21 +1699,26 @@ module Make(Plugin : PLUGIN)
     | Solver_intf.Eval l ->
       let a = mk_atom st f in
       enqueue_semantic st a l
-    | Solver_intf.Consequence (causes, proof) ->
-      let l = List.rev_map (fun f -> Atom.neg @@ mk_atom st f) causes in
-      if List.exists Atom.is_true l then (
-        invalid_argf "slice.acts_propagate: Consequence should contain only true literals"
-      );
+    | Solver_intf.Consequence mk_expl ->
       let p = mk_atom st f in
       if Atom.is_true p then ()
       else if Atom.is_false p then (
+        let lits, proof = mk_expl() in
+        let l = List.rev_map (fun f -> Atom.neg @@ mk_atom st f) lits in
+        if List.exists Atom.is_true l then (
+          invalid_argf "slice.acts_propagate: Consequence should contain only true literals"
+        );
         let c = Clause.make_removable (p :: l) (Lemma proof) in
         raise_notrace (Th_conflict c)
       ) else (
         insert_var_order st p.var;
-        let level = List.fold_left (fun acc a -> max acc a.var.v_level) 0 l in
-        let c = Clause.make_removable (p :: l) (Lemma proof) in
-        enqueue_bool st p ~level (Bcp c)
+        let c = lazy (
+          let lits, proof = mk_expl () in
+          let l = List.rev_map (fun f -> Atom.neg @@ mk_atom st f) lits in
+          Clause.make_removable (p :: l) (Lemma proof)
+        ) in
+        let level = decision_level st in
+        enqueue_bool st p ~level (Bcp_lazy c)
```

Comment:
There is a missing check that the propagation is indeed legal in this case (i.e. the check that all atoms in `lits` are true).

Reply:
It requires to evaluate the justification, which nullifies the point of making it lazy.

Reply:
We could have a `Consequence` and `Consequence_lazy`, tbh (where the first one would check the lits, then reduce to the second one).

Reply:
The check can occur in the `lazy (...)` block, in which case everything should be fine, no ?